### ISSUE TITLE
Persist models back to JCR

### ISSUE THREAD

**Issue by djessup** [unic/neba#132]
I have a project where some fairly complex object graphs need to be persisted to the JCR.

There will be JCR nodes/Sling resources which get adapted to Java models, modified, then committed back to Sling resources. My preference is to use Neba as the model provider since it's IMO more feature-rich and elegant than Sling Models, Slice, etc. 

I understand there are likely to be some challenges in implementing this kind of feature, in particular `AnnotatedFieldMappers` would need to be able to serialise fields as well as deserialise (map) them, though this is hopefully not an insurmountable task.

I'm curious if you feel that persistence has a place in the Neba framework and if it fits the philosophy/roadmap for the product? I will need to implement a solution for my project anyway, and would be happy to contribute some of that work back to Neba if it's of interest. 

Keen to hear your thoughts.

**Comment by olaf-otto** [comment]
Hi @djessup, first of all, thanks for considering the Neba framework for this! 

Generally, I cheerfully welcome the idea. Persisting the models could enable some cool things (like, validating form data and then persisting it...)

Indeed, I spend some time thinking about this approach as well.  IMO, what would be required is an enrichment of the existing annotations to provide sufficient information for persisting the model and, as you mentioned, an extended AnnotatedFieldMapper for custom annotations.

For instance, a specific annotation could be introduced like so

`@PersistableResourceModel(primaryType=x, mixinTypes=x, resourceType=x, resourceSuperType=x)`

The mapping behavior of this model would equal

`@ResourceModel(types = resourceType)
`
Alternatively, one could also come up with a loose contract and instead require the presence of certain annotations to persist a model, like so:

```
@ResourceModel(types=...)
public class MyModel {
     @PrimaryType
     private String primaryType = "this:is:the:default:when:the:model:is:not:loaded:from:jcr";

     @ResourceType
     private String resourceType = "this/is/the/default/when/the/model/is/not/loaded/from/jcr";

    @ResourceSuperType
    ...

    @MixinTypes
    ...
}
```
With pretty much all annotations being optional for persistable models but the @ResourceType annotation.

So, to summarize, I believe this is very much feasible, but the API design will be the crux. Also, there will be some general design questions to consider -  for instance: Do we need a "PersistenceContext" tracking the relationship between the persistent representations (Resources) and models, to ease modification of existing resources (hello, JPA!)? 

At any rate, I will gladly support and welcome your contribution!

Thanks & regards,
Olaf

**Comment by djessup** [comment]
Thanks Olaf, great to hear there is in-principle support for the idea. 

I agree that solid API design is crucial, as would be performance - particularly when processing a large graph. 

Indeed, one of the questions currently rolling around in my head is what the policy for graph traversal would be during a persist operation. 

For instance, say you have three models:
```
class ModelA {
    @Reference ModelB modelB;
}

class ModelB {
    @Reference ModelC modelC;
}

class ModelC {}
```

now you hydrate `ModelA` from a resource and make some changes:

```
ModelA myModelA  = resource.adaptTo(ModelA.class);
myModelA.setSomeProp("value");
myModelA.getModelB().getModelC().setSomeProp("other value");
```

So the graph is `ModelA -> ModelB -> ModelC` and we've modified `ModelA` and `ModelC`, but not `ModelB`. Now we persist `ModelA`:

```
ModelPersister.persist(myModelA, resource);
```

The data in `myModelA` would be written to `resource`. At some stage the `modelB` property would be inspected, and in an efficient system if no changes existed between the model and resource representations it would be skipped. However `ModelC` is "dirty", so by association `ModelB` is too, however we can only discover this by exploring the tree. Imagine a more complex graph and this quickly becomes an expensive endeavour!

All of this is to say, "what to do?". Some options are:

1. Don't persist graphs, only single models
1. Make traversal configurable (e.g. via an annotation `@PersistStrategy(Strategy.IGNORE)`) and let the develop determine what is appropriate
1. Find a way to efficiently track dirty models/change-sets so traversal becomes unnecessary

Or a combination of these, in fact.

I have quite a few thoughts on details of an implementation. Is this the best place to document/discuss, or is there a better venue like a wiki or something?

Cheers
David.

**Comment by olaf-otto** [comment]
@djessup Personally, I am fine with discussing this in this ticket - but I could also setup a wiki if you would prefer that!

Regarding persisting graphs: I believe we must support persisting entire graphs, that is a key feature from my point of view.  I think we should transparently track the relationship of a model to the resource it was mapped from. This should well be feasible, since all adaptations pass the adapter factory. 

Then, we could use this information to decide whether a resource must be created or updated. 

However,this implies that we also model the respective scope. For HTTP requests, we could use the Open-Session-In-View pattern to open up a persistence context for tracking the relationships. outside of a request, we would however need to come up with a different way to open and close the tracking context... perhaps we cold piggyback on Spring here, like DAO support / @Transactional and the like? We could pick up on some good old ideas here, for instance http://www.citytechinc.com/us/en/blog/2008/12/tutorial_jackrabbit.html. What do you think?

Regarding the smartness of the graph traversal, I am thinking of two things we can leverage: 

1) On a Node level, i.e. relationships between resource models: Persistence only cascades to relationships that are either greedy, or lazy relationships that have been loaded. This way, we also achieve conceptual consistency. Would you thing that leveraging lazy loading would suffice to minimize unnecessary graph traversal?

2) On a property level: I believe we can delegate the decision to update or not to update to the ModifiableValueMap view of the resources (or a wrapper thereof, if it's not smart enough). Since this is a hash map, it should be fairly straightforward to decide whether a value was changed I believe.

Regards,
Olaf

**Comment by djessup** [comment]
I agree that full-graph persistence is key, and tracking the lazy relationships sounds like a promisingly elegant solution to over-enthusiastic persistence, I like that one. I'd want to battle-test it because I expect there could be edge cases where a graph might still end up being handle properly, but certainly it sounds like an approach worth pursuing.

Re the persistence context, am I right in thinking you mean essentially a form of session which tracks which of the lazy relationships have been traversed when the persistence operation occurs? If so would this not also naturally be opened at the time a resource is adapted, then closed when the object is persisted or destroyed? Sorry, I feel like I may have misunderstood the question?

**Comment by olaf-otto** [comment]
Regarding the persistence context, Indeed the adapter factory could be the right place to open the context. However, I am not sure about closing / committing it - resources may be opened by application-scoped resource resolvers and live rather long - also, I am unsure how we could track object destruction. Perhaps the most consistent approach is to scope it to the underlying JCR session by default. This, however, would require that we know about the session life cycle events, e.g. destruction callbacks, which I do not see in the API... 

Perhaps explicit scoping, e.g. with something like @Transactional, and a default open session in view filter would be the feasible approach?

**Comment by djessup** [comment]
I don't have experience with `@Transactional` so I'm struggling a bit to understand exactly how it would apply, and I'm wondering if we might be confusing two problems here?

The first is keeping track of what has changed in a model graph, and this is really tied to the lifecycle of the model instance. This context needs to be initialised at the time the instance is created (i.e. adaptation), reset when the model is committed, and closed when the model is destroyed.

The second issue of transactions I see as separate, and from what I've read of `@Transactional` it looks like it could be a useful framework for handling the writeback of models to the JCR in an atomic manner, but to do so efficiently relies on solving the first problem.

So back to problem 1), I have a few ideas for how this could be achieved.

a) When the model is created (adapted) keep a [WeakReference](https://docs.oracle.com/javase/7/docs/api/java/lang/ref/WeakReference.html) to it and when the model becomes unreachable destroy the associated context. 

b) Poll the resolver that provided the resource and if it is no longer live then destroy the context. This makes the assumption that models become invalid once their underlying resource session ends, which seems to make sense to me (and also underlies annotations such as `@This`), but needn't necessarily be the case since a separate resolver could be used for persistence, and in some cases may be desirable.

c) Impose a hard timeout and destroy context sessions that have been open longer than a predefined period.

With a) the risk is that poorly designed consumer code may hold a model reference indefinitely, preventing garbage collection. With b) there is also the risk of long-running sessions, however this is already an accepted risk associated with `ResourceResolver`s. With c) it's clearly a clumsy approach and I'd consider it a last resort. In reality I don't think there is any specific harm keeping these context sessions longer than necessary, aside from the memory they consume.

Any thoughts on any or all of these approaches to handling the context session?

PS. If this is feasible, it may also be possible to implement the standard bean destruction methods like `@PreDestroy`!

**Comment by djessup** [comment]
One further idea I forgot to mention, is it possible to use the lazy loading proxy instance itself to store and query the context, thereby implicitly scoping it to the bean instance without having to actually track the beans lifecycle externally?

**Comment by DaniRey** [comment]
Hi @djessup 

@olaf-otto is very busy this week. As I am not sure what exactly he had in mind, I have a hard time answering your questions. I hope you're ok with receiving an answer from Olaf next week.

Cheers, Dani

**Comment by djessup** [comment]
Hi @DaniRey, no problem at all - I know the feeling!

**Comment by olaf-otto** [comment]
Hi @djessup , I'm finally out of the woods again :-)

First of all, thanks a lot for keeping up the good discussion! I absolutely agree with your list of options and with the pros/cons you mentioned. If we go with an automated context management, I'd probably favor a), as it imposes no additional constraints on the application. However, as you mentioned, all options have specific drawbacks.

What I meant to say in my previous comment regarding @Transactional is that we could instead make the context life cycle explicit, i.e. delegate it's control to the users and expose means to open and close the context.

As a default, we could automatically open and close it using a servlet filter for everything request-scoped, which should cover 95% of all use cases and also matches the scope of the corresponding resource resolver.

For adaptations occurring outside of the request, user's would have to explicitly open and close the context - in case they want to write to the JCR. 

What do you think?

Cheers,
Olaf

**Comment by olaf-otto** [comment]
Hi @djessup!

First of all, happy new year! 
Are you still interested in developing this addition?

Kind regards,
Olaf

**Comment by djessup** [comment]
Hi @olaf-otto, 

And a happy new year to you too, thanks!

Yes, in principle I'm still interested - however I'm in the middle of a career transition and will be travelling a bit the next month. I soon should have a better idea of how much I can commit to working on this - apologies for dragging it out but I'm afraid that's the situation I'm currently in.

I'll try to start getting my brain into gear for this over the next few weeks and will check in mid-Feb with a clearer idea of if and how I might be able to proceed.

Cheers,
David.

**Comment by olaf-otto** [comment]
Hi @djessup 

I will close this issue as won't fix for now - to be re-opened when we find the time to address it!

Kind regards,
Olaf
